### PR TITLE
Split the stack (and many improvements)

### LINKED ISSUE #33: Separate rawrtc.h into a private and a public header
Maybe `rawrtc.h` and `rawrtc_internal.h` or even more fine granular.

### LINKED ISSUE #47: Handle large messages on SCTP redirect transport
The implementation has a fixed size buffer and will not handle large messages atm.

Note: This has nothing to do with the SCTP transport.

### LINKED ISSUE #64: ICE checklist is considered complete too early
`trice_checklist_iscompleted` returns `true` but there are further calls to `ice_failed_handler`. However, the checklist has been only started once. This results in the ICE transport state being set to `COMPLETED` or `FAILED` too early.

This could potentially be a bug in `rew`.

Will post a complete log and maybe info on how to reproduce this issue.

### LINKED ISSUE #82: ICE transport "recovers" from failed state
The ICE transport goes into the `failed` state very early for some candidates and then recovers from it. This shouldn't happen (`failed` is a final state). (And also, it shouldn't go into the `failed` state that early.)

Reproducable in the WebRTC branch with Firefox taking the answering role.

### LINKED ISSUE #104: Add gathering timeout
Sometimes, the ICE gatherer doesn't stop gathering. Although this might be a bug on itself (possible already resolved by the code base of #69), the user application should be able to configure a timeout.

### PR DESCRIPTION
**Pull request by lgrahl**
This splits the stack into three separate entities:
- [RAWRTC common code](https://github.com/rawrtc/rawrtc-common)
- [RAWRTC data channels](https://github.com/rawrtc/rawrtc-data-channel)
- RAWRTC (this library)

This has been done so other libraries can integrate data channels into their WebRTC/ORTC stack without having to use our ICE/DTLS stack.

Note that the required steps to build this library have not been changed (although the `make-dependencies.sh` has become even uglier). However, there will be a subsequent patch addressing this by replacing CMake with consistent usage of Meson across all dependencies making the build process a breeze.

And meanwhile a ton of bug fixes have been applied:
- Resolves #104 
- Resolves #82 
- Resolves #64 
- Resolves #47 
- Resolves #33 

To do:
- [x] Rebase early *WIP* commit messages
- [x] Test ARM detection for disabling hardware CRC32C
- [x] ~~Add an option to turn off hardware CRC32C~~ (Not needed, feature detection has been added to rawrtcdc)
- [x] ~~Ensure it doesn't break worker thread compatibility (#112)~~ It does break worker thread compatibility but we don't consider it an issue. Readme has been updated.
- [x] Update readme
  - We need CMake, Meson and Ninja for now

### CODE REVIEW

**Comment by JesApp** [comment]
This doesn't compile for ARM. See [make-dependencies.log](https://github.com/rawrtc/rawrtc/files/1846885/make-dependencies.log) for what happens when I run `make-dependencies.sh`.

Unless you have implemented some mechanism for detecting ARM builds this is no surprise. We solved that problem last time by simply not building the sctp-redirect-transport by default. You are now using crc32 not just there but in the main rawrtc if I understand the log correctly.

**Comment by lgrahl** [comment]
@JesApp I'm using this to detect ARM now: https://github.com/rawrtc/rawrtc-data-channel/blob/master/src/librawrtcdc/CMakeLists.txt#L44:L47

Can you add the following line (above or below) and let me know what it detects in your case?

```
message(WARNING "processor: ${CMAKE_SYSTEM_PROCESSOR}")
```

**Comment by JesApp** [comment]
Finally had the time to do this, sorry about the delay. The message comes out as `processor: x86_64`. So apparently this detects the processor of the build system, not the target system. I assume this would work if I were building on an ARM processor, instead of cross-compiling.

The solution would be to set `CMAKE_SYSTEM_PROCESSOR` to "arm", either on the command line or in a toolchain file (see [here](https://cmake.org/Wiki/CMake_Cross_Compiling#The_toolchain_file)). But of course that would require editing and re-editing your `make-dependencies.sh` script every time I check out rawrtc.

I remember running into this same problem trying to set the compilers when I was trying to cross-compile rawrtc for the first time. Back then the correct solution would have been a toolchain file as well, but I solved it by setting the `CC` environment variable instead. That won't work for `CMAKE_SYSTEM_PROCESSOR` as far as I know, because cmake doesn't check for an environment variable there.

Basically, this will all go away after #101 is implemented. Unfortunately, I don't have the time to do that right now.

**Comment by lgrahl** [comment]
Ok but it sounds as if this is the correct approach. I'll add an option to turn off hardware CRC32C, so you can turn it off easily.

**Comment by lgrahl** [comment]
Finally picking this up again. :tada:
